src/librustc_mir/build/expr/as_temp.rs

```diff
@@ -26,28 +26,48 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
         where M: Mirror<'tcx, Output = Expr<'tcx>>
     {
         let expr = self.hir.mirror(expr);
-        self.expr_as_temp(block, temp_lifetime, expr)
+        self.expr_as_temp(block, temp_lifetime, expr, false)
+    }
+
+    pub fn as_immutable_temp<M>(&mut self,
+                      block: BasicBlock,
+                      temp_lifetime: Option<region::Scope>,
+                      expr: M)
+                      -> BlockAnd<Local>
+        where M: Mirror<'tcx, Output = Expr<'tcx>>
+    {
+        let expr = self.hir.mirror(expr);
+        self.expr_as_temp(block, temp_lifetime, expr, true)
     }
 
     fn expr_as_temp(&mut self,
                     mut block: BasicBlock,
                     temp_lifetime: Option<region::Scope>,
-                    expr: Expr<'tcx>)
+                    expr: Expr<'tcx>,
+                    immutable: bool)
                     -> BlockAnd<Local> {
-        debug!("expr_as_temp(block={:?}, temp_lifetime={:?}, expr={:?})",
-               block, temp_lifetime, expr);
+        debug!("expr_as_temp(block={:?}, temp_lifetime={:?}, expr={:?}, immutable={})",
+               block, temp_lifetime, expr, immutable);
         let this = self;
 
         let expr_span = expr.span;
         let source_info = this.source_info(expr_span);
         if let ExprKind::Scope { region_scope, lint_level, value } = expr.kind {
             return this.in_scope((region_scope, source_info), lint_level, block, |this| {
-                this.as_temp(block, temp_lifetime, value)
+                if immutable {
+                    this.as_immutable_temp(block, temp_lifetime, value)
```

Comment:
then we would pass the argument here (`this.as_temp(block, temp_lifetime, value, mutability)`), perhaps?